expression/visitor.go

```diff
@@ -0,0 +1,396 @@
+	VisitSubQuery(sq SubQuery) (Expression, error)
+
+	// VisitUnaryOperation visits UnaryOperation expression.
+	VisitUnaryOperation(u *UnaryOperation) (Expression, error)
+
+	// VisitValue visits Value expression.
+	VisitValue(v Value) (Expression, error)
+
+	// VisitValues visits Values expression.
+	VisitValues(v *Values) (Expression, error)
+
+	// VisitVariable visits Variable expression.
+	VisitVariable(v *Variable) (Expression, error)
+
+	// VisitWhenClause visits WhenClause expression.
+	VisitWhenClause(w *WhenClause) (Expression, error)
+}
+
+// BaseVisitor is the base implementation of Visitor.
+// It traverse the expression tree and call expression's Accept function.
+// It can not be used directly.
+// A specific Visitor implementation can embed it in and only implement
+// desired methods to do the job.
+type BaseVisitor struct {
+	V Visitor
+}
+
+// VisitBetween implements Visitor interface.
+func (bv *BaseVisitor) VisitBetween(b *Between) (Expression, error) {
+	var err error
+	b.Expr, err = b.Expr.Accept(bv.V)
+	if err != nil {
+		return b, errors.Trace(err)
+	}
+	b.Left, err = b.Left.Accept(bv.V)
+	if err != nil {
+		return b, errors.Trace(err)
+	}
+	b.Right, err = b.Right.Accept(bv.V)
+	if err != nil {
+		return b, errors.Trace(err)
+	}
+	return b, nil
+}
+
+// VisitBinaryOperation implements Visitor interface.
+func (bv *BaseVisitor) VisitBinaryOperation(o *BinaryOperation) (Expression, error) {
+	var err error
+	o.L, err = o.L.Accept(bv.V)
+	if err != nil {
+		return o, errors.Trace(err)
+	}
+	o.R, err = o.R.Accept(bv.V)
+	if err != nil {
+		return o, errors.Trace(err)
+	}
+	return o, nil
+}
+
+// VisitCall implements Visitor interface.
+func (bv *BaseVisitor) VisitCall(c *Call) (Expression, error) {
+	var err error
+	for i := range c.Args {
+		c.Args[i], err = c.Args[i].Accept(bv.V)
+		if err != nil {
+			return c, errors.Trace(err)
+		}
+	}
+	return c, nil
+}
+
+// VisitCompareSubQuery implements Visitor interface.
+func (bv *BaseVisitor) VisitCompareSubQuery(cs *CompareSubQuery) (Expression, error) {
+	var err error
+	cs.L, err = cs.L.Accept(bv.V)
+	if err != nil {
+		return cs, errors.Trace(err)
+	}
+	_, err = cs.R.Accept(bv.V)
+	if err != nil {
+		return cs, errors.Trace(err)
+	}
+	return cs, nil
+}
+
+// VisitDefault implements Visitor interface.
+func (bv *BaseVisitor) VisitDefault(d *Default) (Expression, error) {
+	return d, nil
+}
+
+// VisitExistsSubQuery implements Visitor interface.
+func (bv *BaseVisitor) VisitExistsSubQuery(es *ExistsSubQuery) (Expression, error) {
+	var err error
+	_, err = es.Sel.Accept(bv.V)
+	if err != nil {
+		return es, errors.Trace(err)
+	}
+	return es, nil
+}
+
+// VisitFunctionCase implements Visitor interface.
+func (bv *BaseVisitor) VisitFunctionCase(f *FunctionCase) (Expression, error) {
+	var err error
+	f.Value, err = f.Value.Accept(bv.V)
+	if err != nil {
+		return f, errors.Trace(err)
+	}
+	for i := range f.WhenClauses {
+		_, err = f.WhenClauses[i].Accept(bv.V)
```

Comment:
Accept returns Expression, but WhenClause is of type `*WhenClause`, it can not be assigned directly but since it's a pointer, it can be modified in `Accept` method, we don't need to assign it back.